### ISSUE TITLE
crash when run a simple bench with concurrency 2 for nightawk server

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/nighthawk#808]
*Title*: *crash when run a simple bench with concurrency 2 for nightawk server*

*Description*:

I created a simple benchmark with nightawk to evaluate perf of Envoy. And I and If I run the nightawk server with concurrency 2 and run the tests, the nightawk server will crash.

*Reproduction steps*:

This problem can be simple reproduced by the following commands.

```
git clone https://github.com/wbpcode/envoy-simple-bench.git
cd envoy-simple-bench
git checkout server-crash
docker-compose up --abort-on-container-exit
```

Ref: https://github.com/wbpcode/envoy-simple-bench/tree/server-crash

*Call Stack*:

```
server_1  | [2022-02-17 03:33:25.432][13][critical][backtrace] [external/envoy/source/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0x0
server_1  | [2022-02-17 03:33:25.432][13][critical][backtrace] [external/envoy/source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
server_1  | [2022-02-17 03:33:25.432][13][critical][backtrace] [external/envoy/source/server/backtrace.h:92] Envoy version: 0/1.21.0-dev//RELEASE/BoringSSL
server_1  | [2022-02-17 03:33:25.445][13][critical][backtrace] [external/envoy/source/server/backtrace.h:98] #0: [0x7f19d8f36090]
server_1  | [2022-02-17 03:33:25.497][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #1: Envoy::Http::FilterManager::decodeData() [0x558cd500267e]
server_1  | [2022-02-17 03:33:25.500][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #2: Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeData() [0x558cd4ff036b]
server_1  | [2022-02-17 03:33:25.504][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #3: Envoy::Http::Http1::ServerConnectionImpl::onMessageCompleteBase() [0x558cd503c0cc]
server_1  | [2022-02-17 03:33:25.507][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #4: Envoy::Http::Http1::ConnectionImpl::onMessageComplete() [0x558cd5039476]
server_1  | [2022-02-17 03:33:25.510][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #5: Envoy::Http::Http1::LegacyHttpParserImpl::Impl::Impl()::{lambda()#3}::__invoke() [0x558cd504338f]
server_1  | [2022-02-17 03:33:25.513][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #6: http_parser_execute [0x558cd5408a39]
server_1  | [2022-02-17 03:33:25.517][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #7: Envoy::Http::Http1::LegacyHttpParserImpl::execute() [0x558cd5042cef]
server_1  | [2022-02-17 03:33:25.520][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #8: Envoy::Http::Http1::ConnectionImpl::dispatchSlice() [0x558cd5037564]
server_1  | [2022-02-17 03:33:25.523][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #9: Envoy::Http::Http1::ConnectionImpl::dispatch() [0x558cd5036cdc]
server_1  | [2022-02-17 03:33:25.526][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #10: Envoy::Http::Http1::ConnectionImpl::dispatch() [0x558cd5037785]
server_1  | [2022-02-17 03:33:25.529][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #11: Envoy::Http::ConnectionManagerImpl::onData() [0x558cd4fea95c]
server_1  | [2022-02-17 03:33:25.538][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #12: Envoy::Network::FilterManagerImpl::onContinueReading() [0x558cd5305c6e]
server_1  | [2022-02-17 03:33:25.542][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #13: Envoy::Network::ConnectionImpl::onReadReady() [0x558cd52fe30f]
server_1  | [2022-02-17 03:33:25.545][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #14: Envoy::Network::ConnectionImpl::onFileEvent() [0x558cd52fb80a]
server_1  | [2022-02-17 03:33:25.554][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #15: std::_Function_handler<>::_M_invoke() [0x558cd52f057f]
server_1  | [2022-02-17 03:33:25.558][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #16: Envoy::Event::FileEventImpl::assignEvents()::$_1::__invoke() [0x558cd52f1892]
server_1  | [2022-02-17 03:33:25.561][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #17: event_process_active_single_queue [0x558cd5427700]
server_1  | [2022-02-17 03:33:25.570][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #18: event_base_loop [0x558cd5426111]
server_1  | [2022-02-17 03:33:25.576][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #19: Envoy::Server::WorkerImpl::threadRoutine() [0x558cd4bd2f36]
server_1  | [2022-02-17 03:33:25.579][13][critical][backtrace] [external/envoy/source/server/backtrace.h:96] #20: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::__invoke() [0x558cd55f97e5]
server_1  | [2022-02-17 03:33:25.579][13][critical][backtrace] [external/envoy/source/server/backtrace.h:98] #21: [0x7f19d8f2b426]
```

**Comment by mum4k** [comment]
Thank you for reporting the crash @wbpcode. As you likely know from the call stack, Nighthawk reuses Envoy's code when issuing requests and receiving responses. This seems to be a segmentation fault in the Envoy code which could either be that this discovered a bug in the Envoy code, or a bug in how Nighthawk uses this portion of the Envoy code.

Would you be interested in (do you have the cycles to) dig deeper into the affected Envoy code and see what condition causes this Segmentation fault? Alternatively I can schedule this work from our end. Let us know.

**Comment by wbpcode** [comment]
I can try to find the problem and fix it. However, I have a lot of work on my own recently, so there may be some delays.

**Comment by mum4k** [comment]
Thank you @wbpcode, feel free to let me know if you will need any help, or if the situation changes.

**Comment by wbpcode** [comment]
This should be fixed by the #809

**Comment by mum4k** [comment]
Thank you for your help @wbpcode!
